sql/analyzer/apply_indexes_for_subquery_comparisons.go

```diff
@@ -31,52 +72,110 @@ import (
-		switch node := node.(type) {
-		case *plan.Filter:
-			var replacement sql.Node
-			if eq, isEqual := node.Expression.(*expression.Equals); isEqual {
-				replacement = getIndexedInSubqueryFilter(ctx, eq.Left(), eq.Right(), node, true, scope, aliases)
-			} else if is, isInSubquery := node.Expression.(*plan.InSubquery); isInSubquery {
-				replacement = getIndexedInSubqueryFilter(ctx, is.Left, is.Right, node, false, scope, aliases)
+		if alias != "" {
+			err = inScope.add(alias, target)
+			if err != nil {
+				return n, transform.SameTree, err
+			}
+		}
+
+		_, isOp := n.(sql.OpaqueNode)
+		outScope := inScope
+		if isOp {
+			outScope = inScope.new()
+		}
+
+		children := n.Children()
+		var newChildren []sql.Node
+		for i, c := range children {
+			child, same, _ := recurse(c, outScope)
+			if !same {
+				if newChildren == nil {
+					newChildren = make([]sql.Node, len(children))
+					copy(newChildren, children)
+				}
+				newChildren[i] = child
 			}
-			if replacement != nil {
-				return replacement, transform.NewTree, nil
+		}
+
+		ret := n
+		same := transform.SameTree
+		if len(newChildren) > 0 {
+			ret, _ = ret.WithChildren(newChildren...)
+			same = transform.NewTree
+		}
+
+		var replacement sql.Node
+		switch n := ret.(type) {
+		case *plan.Filter:
+			switch e := n.Expression.(type) {
+			case *expression.Equals:
+				if rt, ok := n.Child.(*plan.ResolvedTable); ok {
+					replacement = getIndexedInSubqueryFilter(ctx, e.Left(), e.Right(), rt, true, inScope)
+				}
+			case *plan.InSubquery:
+				if rt, ok := n.Child.(*plan.ResolvedTable); ok {
+					replacement = getIndexedInSubqueryFilter(ctx, e.Left, e.Right, rt, false, inScope)
+				}
+			default:
 			}
+		default:
+		}
+		if replacement != nil {
+			ret = replacement
+			same = transform.NewTree
+		}
+
+		if same {
+			return n, transform.SameTree, nil
 		}
-		return node, transform.SameTree, nil
-	})
+		return ret, transform.NewTree, nil
+	}
+	return recurse(n, newScope())
 }
 
 func getIndexedInSubqueryFilter(
 	ctx *sql.Context,
 	left, right sql.Expression,
-	node *plan.Filter,
+	rt *plan.ResolvedTable,
 	equals bool,
-	scope *Scope,
-	tableAliases TableAliases,
+	scope *aliasScope,
 ) sql.Node {
 	gf, isGetField := left.(*expression.GetField)
 	subq, isSubquery := right.(*plan.Subquery)
-	rt, isResolved := node.Child.(*plan.ResolvedTable)
-	if !isGetField || !isSubquery || !isResolved {
-		return nil
-	}
-	referencesChildRow := nodeHasGetFieldReferenceBetween(subq.Query, len(scope.Schema()), len(scope.Schema())+len(node.Child.Schema()))
-	if referencesChildRow {
+	if !isGetField || !isSubquery || !subq.CanCacheResults() {
 		return nil
 	}
 	indexes, err := newIndexAnalyzerForNode(ctx, rt)
 	if err != nil {
 		return nil
 	}
 	defer indexes.releaseUsedIndexes()
-	idx := indexes.MatchingIndex(ctx, ctx.GetCurrentDatabase(), rt.Name(), normalizeExpressions(tableAliases, gf)...)
+	if rt := scope.get(strings.ToLower(gf.Table())); rt != "" {
```

Comment:
(super minor) might be worth adding the `strings.ToLower` to the `aliasScope.get` method so that it's symmetric with how the `aliasScope.add` automatically lower cases the name.

Reply:
yep